### ISSUE TITLE
ssh-copy-id fails if powershell is default shell

### ISSUE THREAD

**Issue by SteveL-MSFT** [PowerShell/PowerShell#4691]
ssh-copy-id is a popular shebang bash script to make it easy to deploy public keys on target machines for key based authn.  
The problem is that this script uses the POSIX shell cmd `exec` which is unknown to powershell:

```bash
ssh "$@" "exec sh -c 'cd ; umask 077 ; mkdir -p .ssh && cat >> .ssh/authorized_keys
```

This can block adoption of powershell as the default shell (via chsh)

Steps to reproduce
------------------

```powershell
PS> ssh-copy-id user@host
```

Expected behavior
-----------------

```none
Success
```

Actual behavior
---------------

```none
ssh-copy-id user@host/usr/bin/ssh-copy-id: INFO: attempting to log in with the new key(s), to filter out any that are already installed
/usr/bin/ssh-copy-id: INFO: 1 key(s) remain to be installed -- if you are prompted now it is to install the new keys
user@host's password: 
exec : The term 'exec' is not recognized as the name of a cmdlet, function, 
script file, or operable program. Check the spelling of the name, or if a path 
was included, verify that the path is correct and try again.
At line:1 char:1
+ exec sh -c 'cd ; umask 077 ; mkdir -p .ssh && cat >> .ssh/authorized_ ...
+ ~~~~
    + CategoryInfo          : ObjectNotFound: (exec:String) [], CommandNotFoun 
   dException
    + FullyQualifiedErrorId : CommandNotFoundException
```

Environment data
----------------

```powershell
Name                           Value                                                                                                
----                           -----                                                                                                
PSVersion                      6.0.0-beta                                                                                           
PSEdition                      Core                                                                                                 
GitCommitId                    v6.0.0-beta.6                                                                                        
OS                             Linux 4.10.0-32-generic #36~16.04.1-Ubuntu SMP Wed Aug 9 09:19:02 UTC 2017                           
Platform                       Unix                                                                                                 
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                              
PSRemotingProtocolVersion      2.3                                                                                                  
SerializationVersion           1.1.0.1                                                                                              
WSManStackVersion              3.0             

```

**Comment by mklement0** [comment]
Typo: The script's name is `ssh-copy-id` - the title and the first sentence of the original post got that wrong.

Somewhat ironically, the `ssh-copy-id` source code states:

> to defend against quirky remote shells: use 'exec sh -c' to get POSIX;

The problem, of course, is that the use of [`exec`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#exec) already presupposes a POSIX-compatible shell.

A quick, but somewhat hacky workaround would be to define `exec` as a built-in alias of `&` - while `&` is _loosely_ the equivalent of POSIX `exec`, there's a crucial difference: the latter _replaces_ the current process with the invoked command's.

**Comment by argonym** [comment]
I am experiencing this interop problem as well with PS 6.1.0 as default shell and OpenSSH 7.8p1 with ProxyJump configured in ~/.ssh/config:
```
PS> ssh -v somehost
debug1: Executing proxy command: exec /usr/bin/ssh [...]
debug1: ssh_exchange_identification: exec : The term 'exec' is not recognized as the name of a cmdlet, function, script file, or operable program.
debug1: ssh_exchange_identification: + exec /usr/bin/ssh [...]
ssh_exchange_identification: Connection closed by remote host [ssh's message is wrong - there has never been any connection]
```
OpenSSH had introduced `exec` as some workaround for Solaris: https://bugzilla.mindrot.org/show_bug.cgi?id=223
Meanwhile they have removed it again for the "match exec" feature: https://bugzilla.mindrot.org/show_bug.cgi?id=2471
However, they didn't remove `exec` for ProxyCommand/ProxyJump even though it seems to be unnecessary there as well meanwhile.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee should review this again.  It seems many unix utils expect the default shell to support `exec`.  For example, I can't use `gdb` if pwsh is the default shell as it tries to call `exec` to run the program under the debugger.

**Comment by joeyaiello** [comment]
I figure this is a pretty inexpensive implementation (either an alias or a thin function wrapper) with the bulk of the difficulty around managing folks who already have a self-defined `exec` in their own environment.

I propose one of two paths:
1. Make this change ASAP in 7.1-preview.1 as an experimental feature. Find out if it breaks anyone long before we deliver it into a supported release.
1. Hardcode a warning for anyone running their own `exec` function or alias into 7.1 previews, with the promise that we'll break them in a subsequent release.

I'm in favor of the first one, personally. Of course, we could also do nothing here, but I have a feeling that this one goes a lot further than `ssh-copy-id` et al

**Comment by SteveL-MSFT** [comment]
PowerShell needs a built-in `exec` that takes the same command line args for compatibility with Unix tooling that expects it so it's a bit more than just an alias to `&`.  Agree that we should have this in early in 7.1 to gauge impact.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we don't believe a simple alias will be sufficient and should look into an actual built-in that calls exec API and only on non-Windows (as the process semantics don't apply to Windows).  Recommend having the same args as bash exec for compatibility.  Get it in early in 7.1 to get feedback and assess user impact due to possibility someone has a `exec` function already defined.

**Comment by brantb** [comment]
`ssh` also tries to call `exec` when using `-J` or the `ProxyJump` directive, which fails when pwsh is the default login shell:

```
PS> ssh target.host.example.com -J jumpbox -v
OpenSSH_8.1p1, LibreSSL 2.7.3
<snip>
debug1: Setting implicit ProxyCommand from ProxyJump: /usr/bin/ssh -v -W '[%h]:%p' jumpbox
debug1: Executing proxy command: exec /usr/bin/ssh -v -W '[target.host.example.com]:22' jumpbox
debug1: kex_exchange_identification: banner line 0: \033[91mexec: \033[91mThe term 'exec' is not recognized as a name of a cmdlet, function, script file, or executable program.
debug1: kex_exchange_identification: banner line 1: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.\033[0m
kex_exchange_identification: Connection closed by remote host

PS> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.1.4
PSEdition                      Core
GitCommitId                    7.1.4
OS                             Darwin 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.…
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

Related: #12750

**Comment by SteveL-MSFT** [comment]
Unfortunately we're not going to get this addressed in 7.2, but we should target a fix in 7.3
